### ISSUE TITLE
Security middleware should patch Manager.raw

### ISSUE THREAD

**Issue by Kazade** [potatolondon/djangae#211]
We should monkey patch it to throw an error, executing raw SQL is insecure. That said it should be disableable, perhaps with a context manager?

**Comment by stucox** [comment]
Maybe just a setting? Which the `check` command we talked about could check is set.

**Comment by Kazade** [comment]
The problem with a setting is it's all or nothing. If the situation arises that raw() needs to be used (if that ever happens...) then the dev will just turn off the setting and then it's pointless. It's better if they have to do something similar to:

```
with yes_i_really_do_need_this():
    Thing.objects.raw("SELECT i_be_crazy from bad_security")
```

**Comment by stucox** [comment]
True. It would still be usable by using `RawQuerySet` directly though, right?

Otherwise Djangae could provide access to the original `raw` method via another name (e.g. `_unsafe_raw`), for those really sure.

The problem with a context manager is it would give the green light to any operations triggered by `Thing.objects.raw()` too (e.g. a signal receiver, etc), or other functions left within the block by a lazy programmer.

I imagine we want to encourage people to wrap any use of raw SQL into reusable, well-tested ORM extensions (as presumably they’re resorting to raw SQL because the ORM doesn’t support their use case). Not sure if the design of this feature might affect whether it’s used like that, but thought I’d say it in case it affects anything

---

*Aside:* While I would totally want this on my projects, I don’t know if Djangae is the place for it: it isn’t an App Engine-specific risk. Maybe it should be in [djangae-scaffold](https://github.com/potatolondon/djangae-scaffold) instead?
